### PR TITLE
[JENKINS-69853] User experimental flags

### PR DESCRIPTION
**Pull request by Wadeck**
See [JENKINS-69853](https://issues.jenkins.io/browse/JENKINS-69853).

This feature proposal adds the possibility to have sort of feature flag per users. The idea behind this is to accelerate the new UI / design work without having to wait for full approval from the community. Publishing something in beta, hidden behind a "user experimental flag" will let the authors to build on top of their previous steps to finalize their feature before the "actual" release.

<details><summary>📷 Screenshots 📷</summary>

---

## With some dummy flags
Screenshot_2022-10-23_004541_001

---

## Without any flags

Screenshot_2022-10-23_005114_001

---

</details>

### Testing done

Local instance running with some dummy flag classes (like you can see in the screenshot).
Also running the test cases to check the different situations.

### Proposed changelog entries

- Introduce user experimental flags.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [n/a] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [n/a] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
 - ⚠️ No extension points so far, the idea is to let the UX SIG team implementing the flags. 

### Desired reviewers

@daniel-beck + @basil for the feature flag topic
@timja + @janfaracik + @NotMyFault for the usage of the feature in UI PRs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7299"></a>

### CODE REVIEW

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: I think https://weekly.ci.jenkins.io/design-library/ToggleSwitch/ would be a more appropriate control

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:54]: Looks a bit weird / overkill would we ever get enough flags to require it? Feature flags should generally be short lived

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:56]: we could probably combine the columns with one of these options:

1. Display name (key)
2. Display name `<span class="jenkins-label jenkins-label--tertiary">key</span>`
image
3. Show it only as a tooltip

Seems a bit unnecessary to display is as the first thing to users

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.properties
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.properties:24]: I think something like one of these, display name doesn't sound great to me as a user facing text for this.

```suggestion
FlagDisplayName=Flag
```

```suggestion
FlagDisplayName=Feature flag
```

```suggestion
FlagDisplayName=Name
```

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:34]: do we need to show this if there isn't any available? (seems unnecessary to me initially as there isn't anything for the user to do here, it's just space taken up for no reason, what do you think?)

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:33]: unsure if this needs to be an alert (if this section stays)

**Review by timja — COMMENTED** [review]
Cool concept

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: The problem with the ToggleSwitch (AFAICT) is that it has only two states.
Having the "default" is important from my PoV so that we can let people decide to "go back" to default.
Huge inspiration from chrome://flags

Screenshot_2022-10-28_181312_001

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:54]: That's a very good point, will be removed. And if at some points we have ton of flags it will be easy to put back :)

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:34]: I thought it has an interest at the beginning or when no flags are active, to just let them know the section exists and it's expected that there is no items inside (could be confusing for users to have nothing inside).

Personally I don't really like the "script security approval" link, that appears only if you have something to approve. That could increase confusion when someone tells you what to look for, or if you want to find back what is the link.

Now, if you think it's better to not have it, I don't have a strong opinion, just personal preference 👍 

Do you think I should make the whole property disabled if there are no flag at the moment? (that's another option as well :D)

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (Wadeck) [inline core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:52]: The `style` is "necessary" to have a space between the two parts. It's the same pattern used in the plugin manager and other location where the `jenkins-label--tertiary` is used (except for the [report issue link](https://github.com/jenkinsci/jenkins/blob/7dd68831c49ca9666ddeaf689cd3c659024005d1/core/src/main/resources/hudson/PluginManager/installed.jelly#L130) in "Installed plugins" tab)

Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:52]: Is there a reason https://weekly.ci.jenkins.io/design-library/Spacing/ can't be used?

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:34]: We can try it out and see if we get any feedback, can always hide it later.

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.properties
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.properties:23]:
```suggestion
True=Enabled
```

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.properties
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.properties:24]:
```suggestion
False=Disabled
```

File: core/src/main/resources/jenkins/model/experimentalflags/Messages.properties
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/Messages.properties:23]: What do you think about just calling this Experiments? i.e. sounds like better user facing terminology to me
```suggestion
UserExperimentalFlagsProperty.DisplayName=Experiments
```

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:33]: What do you think about removing the alert, why do we need to call the users attention to there being no flags?

```suggestion
            <div class="jenkins-form-item">
```

**Review by timja — COMMENTED** [review]
Looking good (tested locally with some flags), do we need any documentation / guidance written for this?

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: What is the default state? (i.e. does this need to be shown visually, in the description or should we simplify with just two states)

File: core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/UserExperimentalFlagsProperty/config.jelly:52]: will have a look 👍 I just re-used the same pattern from other location

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: The default value in the dropdown is "Default". It means that the user does accept the value provided in the code (`getDefaultValue`).

My idea is that we introduce a new feature, we start with disabled by default, so interested users can enable the behavior. During that period, the author can continue improving their feature (without having to wait too much). Then after some months, one can move the feature to be enabled by default. At that moment, the users could decide to rollback by using Disabled. 

In parallel, a user who decide that the feature is interesting they can enabled it in advance and when the feature default behavior will be changed, there is no impact for them.

With the different behavior we can see there, we can gather telemetry to determine if the feature is something the community wants.

I want to be able to see data about adoption instead of just having the opinion from loud users (often either very happy or very unhappy).

Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: I mean how does the user know what the default state is? 'Default' doesn't imply what the default actually is.

Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: oh, in that sense, got it! I will see what I can adjust 👍

File: core/src/main/resources/jenkins/model/experimentalflags/Messages.properties
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/Messages.properties:23]: Moving with `UI Experiments`, is it OK?

_(I cannot find any example of a user feature flag that is not related to UI, as other features are more expected to use SystemProperty for the instance)_

Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/Messages.properties]: why qualify it? seems weird, especially to abbreviate it, thinking of this as a user of Jenkins rather than a developer will they know what UI stands for?

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: Tell me what you think :) 

Screenshot_2022-11-04_121501_001

Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: Yes that's better

Comment (lemeurherve) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: Could the "(default)" be added next to the default entry instead of adding one just for that?

File: core/src/main/resources/jenkins/model/experimentalflags/Messages.properties
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/Messages.properties]: I would think that the qualification help about what it is about.

"User Interface Experiments"?

No strong opinion, just thinking that the "Experiments" could not necessarily convey the meaning.

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: I would say that putting the "help" the closer to where it is used seems better (yeah talking about some pixels :D)

File: core/src/main/resources/jenkins/model/experimentalflags/Messages.properties
Comment (timja) [reply core/src/main/resources/jenkins/model/experimentalflags/Messages.properties]: I wouldn't add too much when we don't really know for sure how it will be used, given there's no PR that's jumping to use this currently.

If you must then I think `User Interface Experiments` is better, but I prefer it without the qualification

**Comment by Wadeck** [comment]
> do we need any documentation / guidance written for this?

TBH I do not know.
It's expected to be used by only a few people at this point, the API is not really complicated, so not sure it's worth the effort.

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (Wadeck) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: After a quick discussion with Hervé (due to my misunderstanding of his argument), I think I forgot to mention the most important argument about the tristate.
As I would like to introduce telemetry on the flags to ease the data gathering, having the "default" option is a way to have a "no vote" / "blank vote" in terms of adoption of the feature.

Comment (lemeurherve) [reply core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:27]: Make perfect sense to have a third default state in that case indeed.

**Comment by NotMyFault** [comment]
> It's expected to be used by only a few people at this point, the API is not really complicated, so not sure it's worth the effort.

We could add a small paragraph to the design library outlining how to use this feature, could be useful I guess.

**Review by uhafner — COMMENTED** [review]
Wouldn't it make more sense to call it feature flag? The idea behind this concept is not only relevant to UI changes.

**Comment by Wadeck** [comment]
> The idea behind this concept is not only relevant to UI changes.

This PR is only about user related flags. I was not able to find a non-UI feature that could be managed by users. For example, notification preference is not a flag, but more a settings. If you have examples of non-UI stuff, please tell me :))

In my mind, the "general" feature flag is more about enabling a particular feature to spread smoothly within an audience.

If you want to use equivalent that are for non-user feature, you already can use the `SystemProperties` (that we heavily use for security corrections with escape hatches)

**Review by timja — APPROVED** [review]
Good for now I think, we can adjust it as needed anyway.

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly
Comment (timja) [inline core/src/main/resources/jenkins/model/experimentalflags/BooleanUserExperimentalFlag/flagConfig.jelly:30]: https://issues.jenkins.io/browse/JENKINS-71211

doesn't fit on mac 13" screen

image

**Pull request merged**
